### PR TITLE
Marblemini [https://github.com/berkeleylab/marble-mini/]

### PR DESCRIPTION
**Pull request by jersey99**
A minimal target.py to follow. @enjoy-digital Let me know if this is a good start.

### CODE REVIEW

File: litex_boards/community/platforms/marblemini.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+# 2. Populate more SFPs
+
+_io = [
+    ("clk20_vcxo", 0, Pins("D17"), IOStandard("LVCMOS33")),
+    ("vcxo_en", 0, Pins("E13"), IOStandard("LVCMOS33")),  # Enable for the above clock
+
+    ("mgt_clk_0", 0,
+     Subsignal("p", Pins("F6")),
+     Subsignal("n", Pins("E6"))),
+    ("mgt_clk_1", 0,
```
Comment (enjoy-digital) [inline litex_boards/community/platforms/marblemini.py]: Can you add an empty lines between items?

File: litex_boards/community/platforms/marblemini.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+    ("ddram",
+        0,
+        Subsignal("a",
+                  Pins("L6 M5 P6 K6 M1 M3 N2 M6", "P1 P2 L4 N5 L3 R1 N3 E3"),
+                  IOStandard("SSTL135")),
+        Subsignal("ba", Pins("L5 M2 N4"), IOStandard("SSTL135")),
+        Subsignal("ras_n", Pins("J4"), IOStandard("SSTL135")),
+        Subsignal("cas_n", Pins("J6"), IOStandard("SSTL135")),
+        Subsignal("we_n", Pins("K3"),  IOStandard("SSTL135")),
+     # Subsignal("cs_n", Pins("T3"), IOStandard("SSTL135")),  # TODO: couldn't find chip select on Marble
```
Comment (enjoy-digital) [inline litex_boards/community/platforms/marblemini.py]: `cs_n` is indeed not mandatory, you can just keep this line commented and remove put an empty `Pins("")`.

File: litex_boards/community/platforms/marblemini.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+        ]
+        self.toolchain.additional_commands = [
+            "write_cfgmem -force -format bin -interface spix4 -size 16 -loadbit \"up 0x0 {build_name}.bit\" -file {build_name}.bin"
+        ]
+        self.add_platform_command("set_property INTERNAL_VREF 0.675 [get_iobanks 35]")
+        self.add_platform_command("set_property CFGBVS VCCO [current_design]")
+        self.add_platform_command("set_property CONFIG_VOLTAGE 3.3 [current_design]")
+
+    def create_programmer(self):
+        # TODO: Should be changed to xc3sprog if confirmed so
```
Comment (enjoy-digital) [inline litex_boards/community/platforms/marblemini.py]: You can also use OpenOCD that is working fine with 7-series (that's what we are using by default on the others platforms).

File: litex_boards/community/platforms/marblemini.py [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+        return VivadoProgrammer()
+
+    def do_finalize(self, fragment):
+        XilinxPlatform.do_finalize(self, fragment)
+        try:
+            self.add_period_constraint(
+                self.lookup_request("clk20_vcxo"), 1e9 / 20e6)
+        except ConstraintError:
+            pass
+        try:
```
Comment (enjoy-digital) [inline litex_boards/community/platforms/marblemini.py]: Now you can just do: `self.add_period_constraint(self.lookup_request("clk20_vcxo", loose=True), 1e9/20e6)` (`loose=True` is there to allow the lookup_request to fail if not found, similarly to what was done before but in a single line).

**Comment by jersey99** [comment]
@enjoy-digital Thanks! Also, thanks for the openocd pointer, the programming works well!

**Comment by enjoy-digital** [comment]
Thanks for the update, merged.

**Comment by enjoy-digital** [comment]
@jersey99: i did some changes to the platform for consistency with the others platforms and to ease maintenance in the future with https://github.com/litex-hub/litex-boards/commit/4faa91c7f28bab3365202db4a59c97c67b5f1a89. Can you check this is fine for you?

**Pull request merged**